src/cell.cpp

```diff
@@ -768,50 +773,88 @@ bool CSGCell::contains_simple(Position r, Direction u, int32_t on_surface) const
 bool CSGCell::contains_complex(
   Position r, Direction u, int32_t on_surface) const
 {
-  // Make a stack of booleans.  We don't know how big it needs to be, but we do
-  // know that rpn.size() is an upper-bound.
-  vector<bool> stack(rpn_.size());
-  int i_stack = -1;
+  // Initialize a stack for operators and the in cell boolean
+  vector<int32_t> op_stack;
+  bool in_cell = false;
 
-  for (int32_t token : rpn_) {
-    // If the token is a binary operator (intersection/union), apply it to
-    // the last two items on the stack. If the token is a unary operator
-    // (complement), apply it to the last item on the stack.
-    if (token == OP_UNION) {
-      stack[i_stack - 1] = stack[i_stack - 1] || stack[i_stack];
-      i_stack--;
-    } else if (token == OP_INTERSECTION) {
-      stack[i_stack - 1] = stack[i_stack - 1] && stack[i_stack];
-      i_stack--;
-    } else if (token == OP_COMPLEMENT) {
-      stack[i_stack] = !stack[i_stack];
-    } else {
-      // If the token is not an operator, evaluate the sense of particle with
-      // respect to the surface and see if the token matches the sense. If the
-      // particle's surface attribute is set and matches the token, that
-      // overrides the determination based on sense().
-      i_stack++;
+  // For each token in prefix
+  for (auto it = region_prefix_.begin(); it != region_prefix_.end(); it++) {
+    // Dereference current iterator
+    int32_t token = *it;
+
+    // If the token is a surface
+    if (token < OP_UNION) {
+      // Evaluate the sense of the particle with respect to the surface and see
+      // if the token matches the sense. If the particle's surface attribute is
+      // set and matches the token, that overrides the determination based on
+      // sense.
       if (token == on_surface) {
-        stack[i_stack] = true;
+        in_cell = true;
       } else if (-token == on_surface) {
-        stack[i_stack] = false;
+        in_cell = false;
       } else {
         // Note the off-by-one indexing
         bool sense = model::surfaces[abs(token) - 1]->sense(r, u);
-        stack[i_stack] = (sense == (token > 0));
+        in_cell = (sense == (token > 0));
       }
+
+      // While the operator stack is populated
+      while (op_stack.size() > 0) {
+        // Extract the top operator
+        int32_t op = op_stack.back();
+        op_stack.pop_back();
+
+        // If the operator is a union and the particle is in the cell or the
+        // the operator is an intersection and the particle is not in the cell
+        if ((op == OP_UNION && in_cell == true) ||
+            (op == OP_INTERSECTION && in_cell == false)) {
+          // Initialize counters for proceeding surfaces and operators
```

Comment:
Here's a suggested explanation of the short circuiting logic we worked through together
```suggestion
          // If we've reached here, we can logically "short circuit" the other
          // operand. To do this, we simply skip over the next branch in the
          // abstract syntax tree. Since union/intersection are binary
          // operators, we need to move forward and count nodes/leaves of the
          // tree until there are one more leaves (surfaces) than nodes
          // (operators).

          // Initialize counters for proceeding surfaces and operators
```